src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,122 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+// The method matches `n` for pattern: AddNode(a, a).
+AddNode::Multiplication AddNode::find_simple_addition_pattern(Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    return Multiplication{true, n->in(1), 2};
+  }
+
+  return Multiplication{};
+}
+
+// Try to match `a << CON`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = 1 << CON`.
+// Match `n` for pattern: LShiftNode(a, CON).
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+AddNode::Multiplication AddNode::find_simple_lshift_pattern(Node* n, BasicType bt) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (!con->is_top()) {
+      return Multiplication{true, n->in(1), java_shift_left(1, con->get_int(), bt)};
+    }
+  }
+
+  return Multiplication{};
+}
+
+// Try to match `CON * a`. On success, return a struct with `.valid = true`, `variable = a`, and `multiplier = CON`.
+// Match `n` for patterns:
+//     - MulNode(CON, a)
+//     - MulNode(a, CON)
+AddNode::Multiplication AddNode::find_simple_multiplication_pattern(Node* n, BasicType bt) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // swap ConNode to lhs for easier matching
+    if (!con->is_Con()) {
+      swap(con, base);
+    }
+
+    if (!con->is_top()) {
+      return Multiplication{true, base, con->get_integer_as_long(bt)};
+    }
+  }
+
+  return Multiplication{};
+}
+
+// Try to match `(a << CON1) + (a << CON2)`. On success, return a struct with `.valid = true`, `variable = a`, and
+// `multiplier = (1 << CON1) + (1 << CON2)`.
+// Match `n` for patterns:
+//     - AddNode(LShiftNode(a, CON), LShiftNode(a, CON)/a)
+//     - AddNode(LShiftNode(a, CON)/a, LShiftNode(a, CON))
+// given that lhs is different from rhs.
+// Note that one of the term of the addition could simply be `a` (i.e., a << 0).
+AddNode::Multiplication AddNode::find_power_of_two_addition_pattern(Node* n, BasicType bt) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) != n->in(2)) {
+    Multiplication lhs = find_simple_lshift_pattern(n->in(1), bt);
+    Multiplication rhs = find_simple_lshift_pattern(n->in(2), bt);
+
+    if (lhs.valid && rhs.valid && lhs.variable == rhs.variable) {
+      return Multiplication{true, lhs.variable, lhs.multiplier + rhs.multiplier};
+    }
+
+    if (lhs.valid && lhs.variable == n->in(2)) {
+      return Multiplication{true, lhs.variable, lhs.multiplier + 1};
+    }
+
+    if (rhs.valid && rhs.variable == n->in(1)) {
+      return Multiplication{true, rhs.variable, rhs.multiplier + 1};
+    }
```

Comment:
Hmm, it seems these are patterns that you did not promise you would cover in the description above.
It makes it a little difficult to keep the overview...

Reply:
Just a drive-by comment what might help: Name the cases you cover in the description with `(1)`, (2)` etc. and add the numbers as comments in the code where you cover the patterns. This would support the mapping from description to implementation.

Reply:
I updated and expanded the comment. They are clearer now. I also added a subcase for `a + a` so `assert()` in `MulNode::Ideal()` can call directly into `find_power_of_two_addition_pattern()`

> Name the cases you cover in the description [..]

That's a great suggestion! Thank you. I'll start using this technique from now on.